### ISSUE TITLE
'System.InvalidOperationException: The reader is closed' when UseDbCallsIfCachingProviderIsDown(true) + not working redis

### ISSUE THREAD

**Issue by azeriker** [VahidN/EFCoreSecondLevelCacheInterceptor#216]
# Summary of the issue
UseDbCallsIfCachingProviderIsDown(true) + not working redis = 

```
System.InvalidOperationException: The reader is closed
   at Npgsql.NpgsqlDataReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDataReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
```

## Environment

```
.NET (Core) SDK version: net7.0
Microsoft.EntityFrameworkCore version: 7.0.11
EFCoreSecondLevelCacheInterceptor version: 6.3.5
Database: PostgreSQL
Cache provider: Redis
```

## Example code/Steps to reproduce:

```
const string CacheProviderName = "main";
services.AddEFSecondLevelCache(options =>
{
    options.UseEasyCachingCoreProvider(CacheProviderName, isHybridCache: false)
        .UseDbCallsIfCachingProviderIsDown(true)
        .DisableLogging(false)
        .UseCacheKeyPrefix("EF_")
        .CacheAllQueriesExceptContainingTableNames(
            CacheExpirationMode.Sliding,
            TimeSpan.FromMinutes(5),
            "__EFMigrationsHistory",
            "pg_class",
            "pg_namespace"
            );
});

services.AddEasyCaching(options =>
{
    const string SerializerName = nameof(SerializerName);
    options.WithJson(SerializerName);

    options.UseRedis(config =>
    {
        config.SerializerName = SerializerName;
        config.DBConfig = new RedisDBOptions
        {
            Configuration = redisConnection
        };
        config.MaxRdSecond = 30;
        config.EnableLogging = true;
        config.LockMs = 5000;
        config.SleepMs = 300;
    }, CacheProviderName);
});
```

## Output:

```
[16:42:36 DBG] Creating DbCommand for 'ExecuteReader'.
[16:42:36 DBG] Creating DbConnection.
[16:42:36 DBG] Created DbConnection. (1ms).
[16:42:36 DBG] Created DbCommand for 'ExecuteReader' (5ms).
[16:42:36 DBG] Initialized DbCommand for 'ExecuteReader' (6ms).
[16:42:36 DBG] Opening connection to database 'dev' on server ''.
[16:42:36 DBG] Opened connection to database 'dev' on server 'tcp://dbhost'.
[16:42:36 DBG] Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='30']
some select command here
[16:42:36 DBG] Using EFCachePolicy: EFCachePolicy --> Sliding|00:05:00|||False.
[16:42:36 DBG] KeyHash: EF_8613880609017F3A, DbContext: EfContext, CacheDependencies: EF_Entities.
[16:42:37 FTL] Interceptor Error
StackExchange.Redis.RedisConnectionException: It was not possible to connect to the redis server(s). Error connecting right now. To allow this multiplexer to continue retrying until it's able to connect, use abortConnect=false in your connection string or AbortOnConnectFail=false; in your code.
   at StackExchange.Redis.ConnectionMultiplexer.ConnectImpl(ConfigurationOptions configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1162
   at StackExchange.Redis.ConnectionMultiplexer.Connect(ConfigurationOptions configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1028
   at StackExchange.Redis.ConnectionMultiplexer.Connect(String configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1002
   at EasyCaching.Redis.RedisDatabaseProvider.CreateConnectionMultiplexer()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at EasyCaching.Redis.RedisDatabaseProvider.GetDatabase()
   at EasyCaching.Redis.DefaultRedisCachingProvider..ctor(String name, IEnumerable`1 dbProviders, IEnumerable`1 serializers, RedisOptions options, RedisLockFactory factory, ILoggerFactory loggerFactory)
   at EasyCaching.Redis.RedisOptionsExtension.<AddServices>b__3_1(IServiceProvider x)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitIEnumerable(IEnumerableCallSite enumerableCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(Type serviceType)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at EFCoreSecondLevelCacheInterceptor.EFEasyCachingCoreProvider.<GetEasyCachingProvider>b__12_0(String name) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\EFEasyCachingCoreProvider.cs:line 213
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at EFCoreSecondLevelCacheInterceptor.EFEasyCachingCoreProvider.GetEasyCachingProvider(EFCacheKey cacheKey) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\EFEasyCachingCoreProvider.cs:line 203
   at EFCoreSecondLevelCacheInterceptor.EFEasyCachingCoreProvider.GetValue(EFCacheKey cacheKey, EFCachePolicy cachePolicy) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\EFEasyCachingCoreProvider.cs:line 122
   at EFCoreSecondLevelCacheInterceptor.DbCommandInterceptorProcessor.ProcessExecutingCommands[T](DbCommand command, DbContext context, T result, EFCachePolicy cachePolicy) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\DbCommandInterceptorProcessor.cs:line 187
   at EFCoreSecondLevelCacheInterceptor.SecondLevelCacheInterceptor.ReaderExecutingAsync(DbCommand command, CommandEventData eventData, InterceptionResult`1 result, CancellationToken cancellationToken) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\SecondLevelCacheInterceptor.cs:line 359
[16:42:38 INF] Executed DbCommand (1,572ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
some select command here
[16:42:38 DBG] Using EFCachePolicy: EFCachePolicy --> Sliding|00:05:00|||False.
[16:42:38 DBG] KeyHash: EF_8613880609017F3A, DbContext: EfContext, CacheDependencies: EF_Entities.
[16:42:38 DBG] Skipped invalidating a none-CRUD command[some select command here].
[16:42:39 FTL] Interceptor Error
StackExchange.Redis.RedisConnectionException: It was not possible to connect to the redis server(s). Error connecting right now. To allow this multiplexer to continue retrying until it's able to connect, use abortConnect=false in your connection string or AbortOnConnectFail=false; in your code.
   at StackExchange.Redis.ConnectionMultiplexer.ConnectImpl(ConfigurationOptions configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1162
   at StackExchange.Redis.ConnectionMultiplexer.Connect(ConfigurationOptions configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1028
   at StackExchange.Redis.ConnectionMultiplexer.Connect(String configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1002
   at EasyCaching.Redis.RedisDatabaseProvider.CreateConnectionMultiplexer()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at EasyCaching.Redis.RedisDatabaseProvider.GetDatabase()
   at EasyCaching.Redis.DefaultRedisCachingProvider..ctor(String name, IEnumerable`1 dbProviders, IEnumerable`1 serializers, RedisOptions options, RedisLockFactory factory, ILoggerFactory loggerFactory)
   at EasyCaching.Redis.RedisOptionsExtension.<AddServices>b__3_1(IServiceProvider x)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitIEnumerable(IEnumerableCallSite enumerableCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(Type serviceType)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at EFCoreSecondLevelCacheInterceptor.EFEasyCachingCoreProvider.<GetEasyCachingProvider>b__12_0(String name) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\EFEasyCachingCoreProvider.cs:line 213
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at EFCoreSecondLevelCacheInterceptor.EFEasyCachingCoreProvider.GetEasyCachingProvider(EFCacheKey cacheKey) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\EFEasyCachingCoreProvider.cs:line 203
   at EFCoreSecondLevelCacheInterceptor.EFEasyCachingCoreProvider.InsertValue(EFCacheKey cacheKey, EFCachedData value, EFCachePolicy cachePolicy) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\EFEasyCachingCoreProvider.cs:line 68
   at EFCoreSecondLevelCacheInterceptor.DbCommandInterceptorProcessor.ProcessExecutedCommands[T](DbCommand command, DbContext context, T result, EFCachePolicy cachePolicy) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\DbCommandInterceptorProcessor.cs:line 126
   at EFCoreSecondLevelCacheInterceptor.SecondLevelCacheInterceptor.ReaderExecutedAsync(DbCommand command, CommandExecutedEventData eventData, DbDataReader result, CancellationToken cancellationToken) in D:\a\EFCoreSecondLevelCacheInterceptor\EFCoreSecondLevelCacheInterceptor\src\EFCoreSecondLevelCacheInterceptor\SecondLevelCacheInterceptor.cs:line 277
[16:42:39 ERR] An exception occurred while iterating over the results of a query for context type 'EfContext'.
System.InvalidOperationException: The reader is closed
   at Npgsql.NpgsqlDataReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDataReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
System.InvalidOperationException: The reader is closed
   at Npgsql.NpgsqlDataReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDataReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
[16:42:40 DBG] Closing data reader to 'dev' on server 'tcp://dbhost'.
[16:42:40 DBG] A data reader for 'dev' on server 'tcp://dbhost' is being disposed after spending 16ms reading results.
[16:42:40 DBG] Closing connection to database 'dev' on server 'tcp://dbhost'.
[16:42:40 DBG] Closed connection to database 'dev' on server '' (2ms).
[16:42:40 INF] Executed action Get (API) in 4089.9201ms
[16:42:40 INF] Executed endpoint 'Get (API)'
[16:42:40 ERR] An unhandled exception has occurred while executing the request.
System.InvalidOperationException: The reader is closed
   at Npgsql.NpgsqlDataReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDataReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
...some my trace
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Hellang.Middleware.ProblemDetails.ProblemDetailsMiddleware.Invoke(HttpContext context)

[16:42:40 INF] Executing ObjectResult, writing value of type 'Microsoft.AspNetCore.Mvc.ProblemDetails'.
[16:42:40 DBG] Connection id "" completed keep alive response.
[16:42:40 DBG] 'EfContext' disposed.
[16:42:40 DBG] Disposing connection to database 'dev' on server ''.
[16:42:40 DBG] Disposed connection to database 'dev' on server '' (2ms).
```

I tried to read your sources, but didnt find any problem.

**Comment by sambar-delen** [comment]
I also encountered this issue. In my issue where I requested the change I continued the conversation (https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/issues/215). I think the problem is that the result get's closed in the try, so when you try and read the result in the catch it has already been closed. (in the_processor.ProcessExecutingCommands(...) function in the SecondLevelCacheInterceptor.cs file).
